src/backends/policy_engine/souffle/check_policy_compliance.cc

```diff
@@ -0,0 +1,100 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+#include <optional>
+#include <sstream>
+
+#include "absl/flags/flag.h"
+#include "absl/flags/parse.h"
+#include "absl/flags/usage.h"
+#include "absl/status/statusor.h"
+#include "src/backends/policy_engine/policy.h"
+#include "src/backends/policy_engine/souffle/souffle_policy_checker.h"
+#include "src/common/logging/logging.h"
+#include "src/parser/ir/ir_parser.h"
+
+ABSL_FLAG(std::string, ir, "", "the IR file");
+ABSL_FLAG(std::string, sql_policy_rules, "",
+          "file containing the SQL policy rules");
+
+constexpr char kUsageMessage[] =
+    "This tool takes an IR representation of a system and returns whether it "
+    "is policy compliant.";
+
+namespace {
+
+absl::StatusOr<std::string> ReadFileContents(std::filesystem::path file_path) {
+  if (!std::filesystem::exists(file_path)) {
+    return absl::NotFoundError(
+        absl::StrCat("File '", file_path.string(), "' does not exist!"));
+  }
+
+  std::ifstream file_stream(file_path);
+  if (!file_stream) {
+    return absl::FailedPreconditionError(absl::StrCat(
+        "Unable to read file '", file_path.string(), "': ", strerror(errno)));
+  }
+
+  std::ostringstream string_stream;
+  string_stream << file_stream.rdbuf();
```

Comment:
I don't think that this is what you want. `rdbuf` just returns the underlying buffer for the file stream; I don't think this is guaranteed to contain the whole file (or, indeed, anything at all).

Reply:
the .str() below converts the entire stream as string?

Reply:
Yes, the `.str()` is correctly converting the contents of the `std::ostringstream string_stream` to a `std::string`, and is capturing the entire contents of `string_stream. My concern, however, is about the line:

```C++
string_stream << file_stream.rdbuf();
```

`file_stream.rdbuf()` just exposes the current contents of the file stream's underlying read buffer. This is just a view into the portion of the file that the stream has currently in memory. It need not be the entire file. Depending on how the filestream is implemented or the conditions on the filesystem when it is opened, I don't think it is even guaranteed to be non-empty for a non-empty file.

Reply:
See version 10 in https://en.cppreference.com/w/cpp/io/basic_ostream/operator_ltlt

Seems like it will do the right thing?

Reply:
Oh, wow, TIL. Thanks for the link, this looks good then.

Reply:
I added a small comment to capture the discussion here.